storagenode/contact/chore.go

```diff
@@ -48,60 +59,106 @@ func (chore *Chore) Run(ctx context.Context) (err error) {
+				}
+				chore.log.Error("ping satellite failed ", zap.Stringer("satellite", satellite), zap.Int("attempts", attempts), zap.Error(err))
+
+				// Sleeps until interval times out, then continue. Returns if context is cancelled.
+				if !sync2.Sleep(ctx, interval) {
+					chore.log.Info("context cancelled", zap.Stringer("satellite", satellite))
+					return nil
+				}
+				interval *= 2
+				if interval >= chore.interval {
+					chore.log.Info("retries timed out for this cycle", zap.Stringer("satellite", satellite))
+					return nil
+				}
+			}
+		})
+	}
+	chore.mu.Unlock()
+	chore.started.Release()
+	return group.Wait()
 }
 
-func (chore *Chore) pingSatellites(ctx context.Context) (err error) {
-	defer mon.Task()(&ctx)(&err)
-	var group errgroup.Group
+func (chore *Chore) pingSatellite(ctx context.Context, id storj.NodeID) (err error) {
+	defer mon.Task()(&ctx, id)(&err)
 	self := chore.service.Local()
-	satellites := chore.trust.GetSatellites(ctx)
+	address, err := chore.trust.GetAddress(ctx, id)
+	if err != nil {
+		return errPingSatellite.Wrap(err)
+	}
+	conn, err := chore.dialer.DialAddressID(ctx, address, id)
+	if err != nil {
+		return errPingSatellite.Wrap(err)
+	}
+	_, err = conn.NodeClient().CheckIn(ctx, &pb.CheckInRequest{
+		Address:  self.Address.GetAddress(),
+		Version:  &self.Version,
+		Capacity: &self.Capacity,
+		Operator: &self.Operator,
+	})
+	if err != nil {
+		return errPingSatellite.Wrap(err)
+	}
+	return nil
+}
 
-	for _, satellite := range satellites {
-		satellite := satellite
-		addr, err := chore.trust.GetAddress(ctx, satellite)
-		if err != nil {
-			chore.log.Error("getting satellite address", zap.Error(err))
-			continue
-		}
-		group.Go(func() error {
-			conn, err := chore.dialer.DialAddressID(ctx, addr, satellite)
-			if err != nil {
-				return err
-			}
-			defer func() { err = errs.Combine(err, conn.Close()) }()
-
-			resp, err := conn.NodeClient().CheckIn(ctx, &pb.CheckInRequest{
-				Address:  self.Address.GetAddress(),
-				Version:  &self.Version,
-				Capacity: &self.Capacity,
-				Operator: &self.Operator,
-			})
-
-			if err != nil {
-				chore.log.Error("Check-In with satellite failed", zap.Error(err))
-			} else if !resp.PingNodeSuccess {
-				chore.log.Error("Check-In with satellite failed due to failed ping back", zap.String("satellite ID", satellite.String()), zap.String("satellite addr", addr), zap.String("ping back error message", resp.GetPingErrorMessage()))
-			}
+// Pause stops all the cycles in the contact chore.
+func (chore *Chore) Pause(ctx context.Context) {
+	chore.started.Wait(ctx)
+	chore.mu.Lock()
+	defer chore.mu.Unlock()
+	for _, cycle := range chore.cycles {
+		cycle.Pause()
+	}
+}
 
-			return err
+// TriggerWait ensures that each cycle is done at least once and waits for completion.
+// If the cycle is currently running it waits for the previous to complete and then runs.
+func (chore *Chore) TriggerWait(ctx context.Context) {
+	chore.started.Wait(ctx)
+	chore.mu.Lock()
```

Comment:
Are these functions going to be able to run if chore.Run still holds the lock?

Reply:
yep it's just locking and unlocking the slice which doesn't take long (the cycles spin up goroutines)